### DISCUSSION TITLE
Support for computed properties

### DISCUSSION THREAD

**Discussion by joao-silveira** [eclipse-langium/langium#1507] [General]
Are there any plans for supporting node properties whose value is computed from other properties in the model?

In my use case I have an interface like:

```
interface INamed {
  name: string;
}
```

And a parsing rule where the name is given either by the referenced node, or by an alias if present:

```
MyRule returns INamedExtended {
   ref=[Target:ID] ('alias' alias=ID)?
}
```

Rough idea would be that we could provide a function that is called when reference linking is done to get the name property from either the reference or the alias property.

**Comment by joao-silveira** [comment]
Another variant of this that i also found that would be useful:

```
interface Direction {
   direction: "NORTH" | "SOUTH"
}
```

```
interface SouthDirection extends Direction {
  // property doesn't need to be parsed and parsing rules can omit it
   direction: "SOUTH"
}
```

**Comment by joao-silveira** [comment]
Just to mention that in Xtext this was supported by `IDerivedStateComputer`. Thanks @cdietrich for the reference.

**Comment by msujew** [comment]
Hey @joao-silveira,

One of Langium's goals was to stay away from the complexity that EMF introduced to Xtext and keep things simple. By now, the core framework is mostly feature complete, and additional functionality (serialization, type systems, etc.) will likely be added as separate packages.

With that in mind, it's unlikely that we plan to support such a feature. Looking at the implementation, you will also see that things work quite a bit differently compared to Xtext, as the whole AST is class-less/prototype-less. We don't plan to introduce more complexity in that region either.

**Reply by cdietrich** [reply]
[@msujew](https://github.com/msujew) we stumbled over this one again. we worked around this by creating artifical node objects inside the scope provider. am not sure if this is a good idea. any recommendations/alternative idea?
